core/src/ledger_cleanup_service.rs

```diff
@@ -62,37 +65,88 @@ impl LedgerCleanupService {
     fn cleanup_ledger(
         new_root_receiver: &Receiver<Slot>,
         blockstore: &Arc<Blockstore>,
-        max_ledger_slots: u64,
-        next_purge_batch: &mut u64,
+        max_ledger_utilization: u64,
+        last_purge_slot: &mut u64,
+        purge_batch_size: u64,
     ) -> Result<(), RecvTimeoutError> {
-        let disk_utilization_pre = blockstore.storage_size();
+        let mut root = new_root_receiver.recv_timeout(Duration::from_secs(1))?;
+        // Get the newest root
+        while let Ok(new_root) = new_root_receiver.try_recv() {
+            root = new_root;
+        }
 
-        let root = new_root_receiver.recv_timeout(Duration::from_secs(1))?;
+        if root - *last_purge_slot > purge_batch_size {
+            *last_purge_slot = root;
 
-        // Notify blockstore of impending purge
-        if root > *next_purge_batch {
-            //cleanup
-            let lowest_slot = root - max_ledger_slots;
-            *blockstore.lowest_cleanup_slot.write().unwrap() = lowest_slot;
-            blockstore.purge_slots(0, Some(lowest_slot));
-            *next_purge_batch += DEFAULT_PURGE_BATCH_SIZE;
-        }
+            debug!("purge: {} last_purge: {}", root, last_purge_slot);
+            let disk_utilization_pre = blockstore.storage_size();
 
-        let disk_utilization_post = blockstore.storage_size();
-
-        if let (Ok(disk_utilization_pre), Ok(disk_utilization_post)) =
-            (disk_utilization_pre, disk_utilization_post)
-        {
-            datapoint_debug!(
-                "ledger_disk_utilization",
-                ("disk_utilization_pre", disk_utilization_pre as i64, i64),
-                ("disk_utilization_post", disk_utilization_post as i64, i64),
-                (
-                    "disk_utilization_delta",
-                    (disk_utilization_pre as i64 - disk_utilization_post as i64),
-                    i64
-                )
+            // Notify blockstore of impending purge
+            //cleanup
+            let mut shreds = Vec::new();
+            let mut iterate_time = Measure::start("iterate_time");
+            let mut total_shreds = 0;
+            for (i, (slot, meta)) in blockstore.slot_meta_iterator(0).unwrap().enumerate() {
+                if i == 0 {
+                    debug!("purge: searching from slot: {}", slot);
+                }
+                total_shreds += meta.completed_data_indexes.len();
```

Comment:
Unfortunately this vector isn't a vector of all the completed shreds, it's a list of the indexes of the "data_complete" shreds. We have no real way to know how many shreds a slot has unless you know it's full `consumed == received`, in which case `received` is the total number of shreds in the slot.

Reply:
The slot should be full since its only rooted slots. I can look at recieved then.

Reply:
There should be roughly equivalent ratio of coding shreds, so it accounts by proxy. Is there an easy way to get number of coding shreds in a slot?